src/NumericalAlgorithms/LinearSolver/Gmres.hpp

```diff
@@ -0,0 +1,209 @@
+          convergence_criteria_, iteration_, initial_residual_magnitude,
+          initial_residual_magnitude);
+      if (convergence_reason_) {
+        break;
+      }
+      operand_ /= initial_residual_magnitude;
+      basis_history_[0] = operand_;
+
+      std::pair<DenseVector<double>, double> minres_and_magnitude{};
+      for (size_t k = 0; k < restart_; k++) {
+        if (use_preconditioner) {
+          preconditioned_basis_history_[k] = preconditioner(operand_);
+        }
+        operand_ = linear_operator(
+            use_preconditioner ? preconditioned_basis_history_[k] : operand_);
+        // Resize matrix
+        orthogonalization_history_.resize(k + 2, k + 1);
+        // Make sure the new entries are zero
+        for (size_t i = 0; i < orthogonalization_history_.rows(); i++) {
+          orthogonalization_history_(
+              i, orthogonalization_history_.columns() - 1) = 0.;
+        }
+        for (size_t j = 0; j < orthogonalization_history_.columns(); j++) {
+          orthogonalization_history_(orthogonalization_history_.rows() - 1, j) =
+              0.;
+        }
+        // Arnoldi orthogonalization
+        for (size_t j = 0; j < k + 1; j++) {
+          const double orthogonalization =
+              inner_product(basis_history_[j], operand_);
+          orthogonalization_history_(j, k) = orthogonalization;
+          operand_ -= orthogonalization * basis_history_[j];
+        }
+        orthogonalization_history_(k + 1, k) =
+            sqrt(inner_product(operand_, operand_));
+        // Least-squares solve for the minimal residual
+        minres_and_magnitude = gmres_detail::minimal_residual(
+            orthogonalization_history_, initial_residual_magnitude);
+        iteration_++;
+        convergence_reason_ = Convergence::criteria_match(
+            convergence_criteria_, iteration_, minres_and_magnitude.second,
+            initial_residual_magnitude);
+        if (UNLIKELY(convergence_reason_)) {
+          break;
+        } else if (k + 1 < restart_) {
+          operand_ /= orthogonalization_history_(k + 1, k);
+          basis_history_[k + 1] = operand_;
+        }
+      }
+      for (size_t i = 0; i < minres_and_magnitude.first.size(); i++) {
+        result += minres_and_magnitude.first[i] *
+                  gsl::at(use_preconditioner ? preconditioned_basis_history_
+                                             : basis_history_,
+                          i);
+      }
+    }
+    return result;
+  }
+
+ private:
+  Convergence::Criteria convergence_criteria_{};
+  ::Verbosity verbosity_{::Verbosity::Verbose};
+  size_t restart_{};
+
+  mutable DenseMatrix<double> orthogonalization_history_{};
+  mutable std::vector<VarsType> basis_history_{};
+  mutable std::vector<VarsType> preconditioned_basis_history_{};
+
+  mutable boost::optional<Convergence::Reason> convergence_reason_{};
+  mutable size_t iteration_{std::numeric_limits<size_t>::max()};
```

Comment:
Add comments about which variables are real state and which are just caches or memory allocations.